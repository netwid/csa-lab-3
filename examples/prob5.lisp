(defun mygcd (a b)
  (if (= b 0)
      a
      (mygcd b (% a b))))

(defun mylcm (a b)
  (/ (* a b) (mygcd a b)))

(defun range (n)
  (if (= n 1)
      1
      (mylcm n (range (- n 1)))))

(range 20)