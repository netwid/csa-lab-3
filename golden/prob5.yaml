source: |-
  (defun mygcd (a b)
    (if (= b 0)
        a
        (mygcd b (% a b))))

  (defun mylcm (a b)
    (/ (* a b) (mygcd a b)))

  (defun range (n)
    (if (= n 1)
        1
        (mylcm n (range (- n 1)))))

  (range 20)

input: |

output: |

code: |-
  [{"index": 0, "opcode": "jmp", "addressing_mode": "direct", "arg": 65, "comment": ""},
   {"index": 1, "opcode": "nop", "addressing_mode": "direct", "arg": 0, "comment": "defun mygcd"},
   {"index": 2, "opcode": "load", "addressing_mode": "immediate", "arg": 0, "comment": ""},
   {"index": 3, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 4, "opcode": "load", "addressing_mode": "stack_relative", "arg": 4, "comment": ""},
   {"index": 5, "opcode": "cmp", "addressing_mode": "stack_relative", "arg": 1, "comment": ""},
   {"index": 6, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 7, "opcode": "je", "addressing_mode": "direct", "arg": 9, "comment": ""},
   {"index": 8, "opcode": "jmp", "addressing_mode": "direct", "arg": 11, "comment": ""},
   {"index": 9, "opcode": "load", "addressing_mode": "stack_relative", "arg": 2, "comment": ""},
   {"index": 10, "opcode": "jmp", "addressing_mode": "direct", "arg": 22, "comment": ""},
   {"index": 11, "opcode": "load", "addressing_mode": "stack_relative", "arg": 3, "comment": ""},
   {"index": 12, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 13, "opcode": "load", "addressing_mode": "stack_relative", "arg": 3, "comment": ""},
   {"index": 14, "opcode": "mod", "addressing_mode": "stack_relative", "arg": 1, "comment": ""},
   {"index": 15, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 16, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 17, "opcode": "load", "addressing_mode": "stack_relative", "arg": 4, "comment": ""},
   {"index": 18, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 19, "opcode": "call", "addressing_mode": "direct", "arg": 1, "comment": ""},
   {"index": 20, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 21, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 22, "opcode": "ret", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 23, "opcode": "nop", "addressing_mode": "direct", "arg": 0, "comment": "defun mylcm"},
   {"index": 24, "opcode": "load", "addressing_mode": "stack_relative", "arg": 3, "comment": ""},
   {"index": 25, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 26, "opcode": "load", "addressing_mode": "stack_relative", "arg": 3, "comment": ""},
   {"index": 27, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 28, "opcode": "call", "addressing_mode": "direct", "arg": 1, "comment": ""},
   {"index": 29, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 30, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 31, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 32, "opcode": "load", "addressing_mode": "stack_relative", "arg": 4, "comment": ""},
   {"index": 33, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 34, "opcode": "load", "addressing_mode": "stack_relative", "arg": 4, "comment": ""},
   {"index": 35, "opcode": "mul", "addressing_mode": "stack_relative", "arg": 1, "comment": ""},
   {"index": 36, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 37, "opcode": "div", "addressing_mode": "stack_relative", "arg": 1, "comment": ""},
   {"index": 38, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 39, "opcode": "ret", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 40, "opcode": "nop", "addressing_mode": "direct", "arg": 0, "comment": "defun range"},
   {"index": 41, "opcode": "load", "addressing_mode": "immediate", "arg": 1, "comment": ""},
   {"index": 42, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 43, "opcode": "load", "addressing_mode": "stack_relative", "arg": 3, "comment": ""},
   {"index": 44, "opcode": "cmp", "addressing_mode": "stack_relative", "arg": 1, "comment": ""},
   {"index": 45, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 46, "opcode": "je", "addressing_mode": "direct", "arg": 48, "comment": ""},
   {"index": 47, "opcode": "jmp", "addressing_mode": "direct", "arg": 50, "comment": ""},
   {"index": 48, "opcode": "load", "addressing_mode": "immediate", "arg": 1, "comment": ""},
   {"index": 49, "opcode": "jmp", "addressing_mode": "direct", "arg": 64, "comment": ""},
   {"index": 50, "opcode": "load", "addressing_mode": "immediate", "arg": 1, "comment": ""},
   {"index": 51, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 52, "opcode": "load", "addressing_mode": "stack_relative", "arg": 3, "comment": ""},
   {"index": 53, "opcode": "sub", "addressing_mode": "stack_relative", "arg": 1, "comment": ""},
   {"index": 54, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 55, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 56, "opcode": "call", "addressing_mode": "direct", "arg": 40, "comment": ""},
   {"index": 57, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 58, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 59, "opcode": "load", "addressing_mode": "stack_relative", "arg": 3, "comment": ""},
   {"index": 60, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 61, "opcode": "call", "addressing_mode": "direct", "arg": 23, "comment": ""},
   {"index": 62, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 63, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 64, "opcode": "ret", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 65, "opcode": "load", "addressing_mode": "immediate", "arg": 20, "comment": ""},
   {"index": 66, "opcode": "push", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 67, "opcode": "call", "addressing_mode": "direct", "arg": 40, "comment": ""},
   {"index": 68, "opcode": "pop", "addressing_mode": "direct", "arg": 0, "comment": ""},
   {"index": 69, "opcode": "halt", "addressing_mode": "direct", "arg": 0, "comment": ""}]

log: []
